### PR TITLE
Changes default DDP behavior to divide sparse grad by world size before allreduce, not after

### PR DESCRIPTION
**Pull request by mcarilli**
I appreciate https://github.com/pytorch/pytorch/pull/61379, which restores the fusion of div-by-world-size and copy-to-allreduce-buffer for dense gradients. But i noticed in the wake of https://github.com/pytorch/pytorch/pull/61379 there's misaligned treatment of dense and sparse gradients. Specifically, dense gradients are dived by world size before the allreduce, and sparse gradients are dived by world size after the allreduce. On paper you wouldn't expect that to matter, but for cluster-scale DDP training with amp gradient scaling and allreduces of FP16 grads, we've noticed several cases where postdividing grads by world size caused nonconvergence while predividing worked. I'm not aware of any cases where the reverse was true.

This PR changes the treatment of sparse gradients to match the treatment of dense gradients (both will be dived by world size before allreduce).

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]
It's very strange that pre vs post dividing will cause convergence issues, do we know why this is? Could it possibly be related to pre-dividing helping prevent overflow that allreduce sum can incur (and if so are we concerned about the case where the overflow is so significant that division by world_size doesn't help)?

**Comment by mcarilli** [comment]
> Could it possibly be related to pre-dividing helping prevent overflow that allreduce sum can incur

Yes. It's an interesting issue. With amp training you never want to unscale gradients while they're still FP16. So for networks set up to allreduce FP16 grads, these grads will be scaled by the amp scale factor. If the allreduce is across many ranks, the allreduce can hit overflows, which drives the scale factor down. If you predivide, you're less vulnerable to this overflow. Put another way, the scale factor within each process has more freedom to dynamically adjust to greater magnitudes (which makes _non-leaf_ grads within each process's backward pass less vulnerable to _underflow_).
